src/Build.UnitTests/WeakStringCache_Tests.cs

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+using Shouldly;
+using Xunit;
+
+namespace Microsoft.Build.UnitTests
+{
+    public class WeakStringCache_Tests : IDisposable
+    {
+        /// <summary>
+        /// The weak string cache under test.
+        /// </summary>
+        private WeakStringCache _cache = new WeakStringCache();
+
+        public void Dispose()
+        {
+            _cache.Dispose();
+        }
+
+        /// <summary>
+        /// Triggers full GC that reliably collects all objects without strong GC roots and also nulls out weak GC handles.
+        /// </summary>
+        /// <remarks>
+        /// While a simple GC.Collect() would be enough when running on .NET Framework or .NET Core, the Mono runtime appears to have
+        /// GC handles implemented differently and requires a bit more effort. We don't care when exactly our handles are nulled out
+        /// as long as it eventually happens for otherwise unreachable strings.
+        /// </remarks>
+        private void RunGC()
```

Comment:
Current mono failures all refer to RunGC, and all tests that refer to it seem to have failed.

Reply:
Aha! I finally figured out why these tests have been problematic on Mac Mono. From https://www.mono-project.com/docs/advanced/garbage-collector/sgen/#precise-stack-marking:

_"Both Boehm GC, and SGen on macOS scan threads stacks conservatively. This means that they treated every value found on a stack as a potential pointer. This is problematic for several reasons:_

_- The possibility that some objects remain alive by mistake, because a value which points to them is found on the stack."_

Conservative GC scanning makes it very hard to write a reliable test where we expect an object to get collected. I could try to hack around it but I think we can just disable these tests on Mono to save us from future pain.

Reply:
Sounds fine to me. Maybe link to that page for posterity :)